### ISSUE TITLE
Math.xxx >50x slower in userscripts

### ISSUE THREAD

**Issue by RubenArtmann** [violentmonkey/violentmonkey#1531]
# Issue
calling `Math.abs`, `Math.random`, etc. is >50x slower in userscripts

## test
to test just copy the following into a userscript open the site it is injected into, and paste the same into the console:
```js
let acc = 0;
console.time("Math.randomx1000000");
for(let i=0; i<1000000; i++) acc += Math.random();
console.timeEnd("Math.randomx1000000");
console.time("Math.absx1000000");
for(let i=0; i<1000000; i++) acc += Math.abs(acc);
console.timeEnd("Math.absx1000000");

// result in userscript:
//   Math.randomx1000000: 4118.552978515625 ms
//   Math.absx1000000: 4082.463134765625 ms
// result in console:
//   Math.randomx1000000: 135.682861328125 ms
//   Math.absx1000000: 55.900634765625 ms
```
## temporary fix
rebinding the object like this fixes the problem:
```js
const Math = window.Math;
```

despite the simple fix, this caused me major issues (freezes when doing image manipulation) and headaches (took me way to long to debug, because the profiler just shows 40s spent with `Math.abs` being the only thing on that line)
## speculation on the origin
I assume the property access somehow has overhead because of sand-boxing or something, ~~but the Math object is indistinguishable from the normal one (no proxy, no funky getters / setters), so I have no clue what is going on~~.

**Comment by tophf** [comment]
This is caused by our implementation of sandboxing. We use `Proxy`, which performs a lookup for each global. Normally, this is completely negligible. The only way to make it faster, I guess, is to get rid of Proxy at the cost of not supporting `window.GMxxx` which would break [hundreds of scripts](https://greasyfork.org/en/scripts/code-search?c=window.GM) though and we won't be able to avoid leaking of undeclared global variables in non-strict mode e.g. `foo = 123` which beginners are very likely to write. So, I'd say this is working-as-intended.

**Comment by RubenArtmann** [comment]
Ah, so its not:
```js
   Math.abs
       ^
```
thats causing the problem, but:
```js
   window.Math.abs
proxy^   ^
```
which means, that
```js
const Math = window.Math;
```
fixes the issue, because you only have to pay for the proxy once. 

Why would that be problematic?

**Comment by tophf** [comment]
> Why would that be problematic? (As long as you put it in an outer scope, so you can shadow it)

You can do it in your code without any problems because `Math` doesn't have need to be bound to `window`. Only [some methods](https://github.com/violentmonkey/violentmonkey/blob/3fe3ec051e739d97213567eb5f41dc462afbe074/src/injected/web/gm-global-wrapper.js#L83) need it.

I don't think we should do it automatically because it will break scripts that replace globals e.g. window.Math = somePolyfill.

> count the number of times the proxy's handlers get triggered

No, because counting itself increases the number of lookups which makes everything even slower.

**Comment by RubenArtmann** [comment]
> I don't think we should do it automatically because it will break scripts that replace globals e.g. window.Math = somePolyfill.

What about this:
```js
const wrapper_template = (()=>{
  const userscriptScope = {
  	GM_MESSAGE: "GM_ apis work"
  };
  const fastPropertyGetters = {};
  const fastPropertySetters = {};
  const windowProxy = new Proxy(window,{
    get: (target,prop,receiver)=>{
//       // this is to much overhead?
//       proxyOverheadCounter++ if(proxyOverheadCounter>10000) console.warn("[Violentmonkey] performance issue due to sandboxing detected (see https://github.com/violentmonkey/violentmonkey/issues/1531)")
      
      if(fastPropertyGetters[prop]) return fastPropertyGetters[prop]();
      let value = Reflect.get(target,prop);
//       userscriptScope[prop] = value; // ~5x acceleration
//       // global window[prop] = value no longer updates userscripts view
      return value;
    },
    set: (target,prop,value,receiver)=>{
      if(fastPropertySetters[prop]) return fastPropertySetters[prop](value);
      return Reflect.set(userscriptScope,prop);
    }
  });
  userscriptScope.window = windowProxy;
  userscriptScope.unsafeWindow = window;
  with(windowProxy) with(userscriptScope) {
    /*FAST_PROPERTIES*/
    /*USER_SCRIPT*/
  }
}).toString().slice(5,-1).replace(/\n\s\s/g,"\n");

const FAST_PROPERTIES = ["Math","Object","Array","document"]; // up to ~500x acceleration
const compileUserscript = (userscript)=>{
  return wrapper_template.replace("/*FAST_PROPERTIES*/",FAST_PROPERTIES.map(prop=>
    `let ${prop} = window.${prop}; fastPropertyGetters[${JSON.stringify(prop)}] = ()=>${prop}; fastPropertySetters[${JSON.stringify(prop)}] = v=>${prop}=v;`
  ).join("\n  ")).replace("/*USER_SCRIPT*/",userscript);
};

const userscript = (()=>{
  // you dont need to break GM_ apis
  console.log(GM_MESSAGE);
  
  // ~500x faster for allowlisted properties
  console.time("Math.abs");
  for(let i=0; i<1000000; i++) Math.abs(i,i+1);
  // with fastprop: ~10ms
  // without: ~5000ms
  console.timeEnd("Math.abs");
  
  // you can polyfill or whatever
  Math = "a";
  console.log(Math,window.Math); // a a
  window.Math = "b";
  console.log(Math,window.Math); // b b
  
  // side-effects:
  //   * scripts cannot polute direct window properties (without using unsafeWindow) (can be changed by manually poluting in windowProxy)
  //   * the site or other scripts cannot polute allowlisted properties (can be changed by hooking using window.__defineSetter__ to call fastPropertySetters[<prop>](value))
}).toString().slice(5,-1).replace(/\n\s*/g,"\n");

let compiled = compileUserscript(userscript);
console.log(compiled);
eval(compiled);
```

**Comment by RubenArtmann** [comment]
> The very fact that there is even a single lookup already incurs the penalty.

no, there is no lookup. it injects `let Math = window.Math` into the userscript, but `window.Math = somePolyfill` still works.

**Comment by RubenArtmann** [comment]
> Looks like the ingenious trick you're using is that the second `with` lets JS itself perform the lookup.

Thats *one* trick, that "only" saves ~5x. 

`with` is still really expensive

the other is injecting `let Math = window.Math` with some trickery directly into the script to save ~500x.

**Comment by tophf** [comment]
I don't think we'll do the second thing because we would need to do it for all globals, not just Math, and there are like 1000 or 2000 of them. We did that in the past, just like Tampermonkey, but copying them all took for every userscript takes a lot of time e.g. 100ms, which is **several times** longer than our current initialization. This is 6 paint frames so it'll cause jank at DOMContentLoaded/load events or it'll delay page load at document_start.

**Comment by tophf** [comment]
Yeah that's sensible.

**Comment by RubenArtmann** [comment]
Just an Idea: you could count how many times each property gets used and then dynamicly enable option 2 for everything where the overhead would be greater than 100ms on the next run.

**Comment by tophf** [comment]
I've implemented both tricks in #1532.

As for adaptive caching, I doubt it's trivial so I'm afraid it'll unnecessarily complicate code.
